source/dem/dem.cc

```diff
@@ -352,18 +352,26 @@ template <int dim>
 void
 DEMSolver<dim>::update_moment_of_inertia(
   dealii::Particles::ParticleHandler<dim> &          particle_handler,
-  std::unordered_map<types::particle_index, double> &MOI)
+  std::vector<double> &MOI)
 {
-  // Clearing the container first
-  MOI.clear();
+  unsigned int max_particle_id = 0;
+  for (const auto &particle : particle_handler)
+    max_particle_id = std::max(max_particle_id, particle.get_id());
+
+  // We resize force and momentum every time we sort the particles
+  // into subdomains
+  momentum.resize(max_particle_id + 1);
+  force.resize(max_particle_id + 1);
+  displacement.resize(max_particle_id + 1);
+  MOI.resize(max_particle_id + 1);
```

Comment:
Is this not dangerous in parallel? For example, if a processor has particle id 8M, won't that reserve an 8M vector?

Reply:
You are right - I misread the deal.II documentation, I thought they were local to each MPI process. But the solution is to make the accessor classes in deal.II aware of an MPI-local index - it does exist because this is what is used for the property pool and other access information inside deal.II, so it is just a matter of finding it (or exposing it in deal.II's classes). Let me have a look at it later today or tomorrow.

Reply:
I had a short look now: I think we would like to get the information inside `get_handle()` as a way to have indexing in MPI-local index space, https://github.com/dealii/dealii/blob/b4e88620cab2396f6daf8d36c78553ecc6a3301d/include/deal.II/particles/particle_accessor.h#L327-L328. We don't want to make that function public, but I think it would make a lot of sense to add a function `get_local_index()` (or `get_local_id()`) to `ParticleAccessor` that returns the number in MPI-local index space. I don't know if we want to make the numbers a tight map in the sense that we keep the numbers contiguous or if we simply use the number that is used inside the property pool. For our purposes, the latter would suffice, as we update it whenever the particles move around. Before going more, two short questions:
- In order to satisfy backward compatibility with Lethe, we will need to make this conditional and have a translation layer for the older version. Any other comments on this topic?
- Do we need to carry over the forces, momentum, etc. when the particle assignments to cells change? I don't think so, because we have https://github.com/lethe-cfd/lethe/blob/59b11a72dfde82d9d1d48b646bf9dfff4c517e93/source/dem/dem.cc#L749 and similar statements for the rest. This would mean we do not need to be particularly careful in keeping the numbering alive from Lethe's end, and we could take whatever numbers deal.II gives us.

With the closer look I realized what confused me to make this wrong assumption, it is the fact that you use `unsigned int particle_one_id = ...` e.g. in https://github.com/lethe-cfd/lethe/blob/59b11a72dfde82d9d1d48b646bf9dfff4c517e93/source/dem/pp_nonlinear_force.cc#L188 which I interpreted as the local index, as it differs from `types::particle_index` in the context I use to work with (big parallel machines with 64 bit indices). In any case, this suggestion was also among the optimizations we introduced when writing the `LinearAlgebra::distributed::Vector` class in deal.II: When we work with some array structures that are located on the current MPI process, we really do not want to go through the index translation from the global space and then somehow back again, but we should immediately use the local one. Of course, the global index is the more convenient one because it helps us to synchronize the view of different MPI processes on the index, but the internal data structures should be designed differently.

Reply:
Please find my comments below

> I had a short look now: I think we would like to get the information inside `get_handle()` as a way to have indexing in MPI-local index space, https://github.com/dealii/dealii/blob/b4e88620cab2396f6daf8d36c78553ecc6a3301d/include/deal.II/particles/particle_accessor.h#L327-L328. We don't want to make that function public, but I think it would make a lot of sense to add a function `get_local_index()` (or `get_local_id()`) to `ParticleAccessor` that returns the number in MPI-local index space. I don't know if we want to make the numbers a tight map in the sense that we keep the numbers contiguous or if we simply use the number that is used inside the property pool. For our purposes, the latter would suffice, as we update it whenever the particles move around. Before going more, two short questions:
> 
> * In order to satisfy backward compatibility with Lethe, we will need to make this conditional and have a translation layer for the older version. Any other comments on this topic?

You mean to ensure backward compatibility with deal.II 9.3 right? In this case yeah, we would need to have a translation layer. I don't like it very much when there are too much #ifdef lying around or when the behavior changes drastically from 9.3 to 10.0pre, but to be honest, there is no way around this. We had the same thing happen to use when we introduced update_ghost_particles. The deal.II particle library is still immature, this is bound to happen.

> * Do we need to carry over the forces, momentum, etc. when the particle assignments to cells change? I don't think so, because we have https://github.com/lethe-cfd/lethe/blob/59b11a72dfde82d9d1d48b646bf9dfff4c517e93/source/dem/dem.cc#L749
>    and similar statements for the rest. This would mean we do not need to be particularly careful in keeping the numbering alive from Lethe's end, and we could take whatever numbers deal.II gives us.
We do not. Maybe @shahabgol can comment on this, but as far as I know, the only thing we need to keep alive and coherent in a contact in pretty much the tangential overlap  (because it carries over in a contact history) and this is kept in a seperate structure which is defined on a id-id basis, the force and the other contact properties do not need to be ept.

> 
> With the closer look I realized what confused me to make this wrong assumption, it is the fact that you use `unsigned int particle_one_id = ...` e.g. in
> 
> https://github.com/lethe-cfd/lethe/blob/59b11a72dfde82d9d1d48b646bf9dfff4c517e93/source/dem/pp_nonlinear_force.cc#L188
> 
> which I interpreted as the local index, as it differs from `types::particle_index` in the context I use to work with (big parallel machines with 64 bit indices). In any case, this suggestion was also among the optimizations we introduced when writing the `LinearAlgebra::distributed::Vector` class in deal.II: When we work with some array structures that are located on the current MPI process, we really do not want to go through the index translation from the global space and then somehow back again, but we should immediately use the local one. Of course, the global index is the more convenient one because it helps us to synchronize the view of different MPI processes on the index, but the internal data structures should be designed differently.

Using unsigned int is a mistake on our part, types::particle_index should have been used from the get go. I fully agree with you that having a local id as a 32bit index would be extremely relevant.  I don't think we need to keep that number contiguous, but having some sort of way indexing the particles "Locally" instead of globally would actually allow us to simplify a lot of the stuff happening because there would be a reliable way to create a vector-like container for the local particles.

Reply:
Sorry that I can only come back to this one now, but I think we should progress now. I will open a PR with a suitable function in deal.II now and then we can see how we can integrate the new functionality in a backward-compatible way.

Reply:
Absolutely. In the worst case, we can make the 9.3 version "backward compatible" without being optimal and provide the correct documentation.

Reply:
> Please find my comments below
> 
> > I had a short look now: I think we would like to get the information inside `get_handle()` as a way to have indexing in MPI-local index space, https://github.com/dealii/dealii/blob/b4e88620cab2396f6daf8d36c78553ecc6a3301d/include/deal.II/particles/particle_accessor.h#L327-L328. We don't want to make that function public, but I think it would make a lot of sense to add a function `get_local_index()` (or `get_local_id()`) to `ParticleAccessor` that returns the number in MPI-local index space. I don't know if we want to make the numbers a tight map in the sense that we keep the numbers contiguous or if we simply use the number that is used inside the property pool. For our purposes, the latter would suffice, as we update it whenever the particles move around. Before going more, two short questions:
> > 
> > * In order to satisfy backward compatibility with Lethe, we will need to make this conditional and have a translation layer for the older version. Any other comments on this topic?
> 
> You mean to ensure backward compatibility with deal.II 9.3 right? In this case yeah, we would need to have a translation layer. I don't like it very much when there are too much #ifdef lying around or when the behavior changes drastically from 9.3 to 10.0pre, but to be honest, there is no way around this. We had the same thing happen to use when we introduced update_ghost_particles. The deal.II particle library is still immature, this is bound to happen.
> 
> > * Do we need to carry over the forces, momentum, etc. when the particle assignments to cells change? I don't think so, because we have https://github.com/lethe-cfd/lethe/blob/59b11a72dfde82d9d1d48b646bf9dfff4c517e93/source/dem/dem.cc#L749
> >   
> >   and similar statements for the rest. This would mean we do not need to be particularly careful in keeping the numbering alive from Lethe's end, and we could take whatever numbers deal.II gives us.
> >   We do not. Maybe @shahabgol can comment on this, but as far as I know, the only thing we need to keep alive and coherent in a contact in pretty much the tangential overlap  (because it carries over in a contact history) and this is kept in a seperate structure which is defined on a id-id basis, the force and the other contact properties do not need to be ept.
> 
> > With the closer look I realized what confused me to make this wrong assumption, it is the fact that you use `unsigned int particle_one_id = ...` e.g. in
> > https://github.com/lethe-cfd/lethe/blob/59b11a72dfde82d9d1d48b646bf9dfff4c517e93/source/dem/pp_nonlinear_force.cc#L188
> > 
> > which I interpreted as the local index, as it differs from `types::particle_index` in the context I use to work with (big parallel machines with 64 bit indices). In any case, this suggestion was also among the optimizations we introduced when writing the `LinearAlgebra::distributed::Vector` class in deal.II: When we work with some array structures that are located on the current MPI process, we really do not want to go through the index translation from the global space and then somehow back again, but we should immediately use the local one. Of course, the global index is the more convenient one because it helps us to synchronize the view of different MPI processes on the index, but the internal data structures should be designed differently.
> 
> Using unsigned int is a mistake on our part, types::particle_index should have been used from the get go. I fully agree with you that having a local id as a 32bit index would be extremely relevant. I don't think we need to keep that number contiguous, but having some sort of way indexing the particles "Locally" instead of globally would actually allow us to simplify a lot of the stuff happening because there would be a reliable way to create a vector-like container for the local particles.

We need the force and momentum in the integration to obtain the acceleration, velocity, position, and angular velocities on the particles. Plus, we need them in the visualization. Other than that, we don't need force and momentum in the next time-step.